### PR TITLE
Opencontrail integration

### PR DESCRIPTION
**Pull request by sanju-a**
# _Opencontrail Network Provider Kubernetes_

The commits in this PR consist of the following:
1- Salt changes to provision opencontrail in master and minion
2- Salt changes to provision opencontrail gateway
3- Opencontrail gateway functionality on a separate VM or on the first minion is environment
variable driven. Default is to deploy on separate VM.
4- Issue with race condition in kube-apiserver and kubelet, docker setup is addressed
5- Default variables for initializing image, compute and memory and disk size is defaulted for opencontrail
6- Debain VM - wheezy 7.9, kernel - 3.16 is used when opencontrail is used as the network provider
7- To initialize opencontrail network provider the following environment variable is required

**export NETWORK_PROVIDER=opencontrail**

### CODE REVIEW

**Comment by davidopp** [comment]
Reassigning to @thockin

**Comment by thockin** [comment]
I am travelling the next 2 weeks and might be slow to respond.  It's in my queue

**Comment by sanabby** [comment]
Sure! Thanks Tim. Will look forward for the review comments.

**Comment by sanabby** [comment]
Tim, do you have some time to look at this review. I can spend sometime to demo it and go over the salt changes. Appreciate if you could let me know.

**Comment by eparis** [comment]
ok to test

File: cluster/gce/config-default.sh [hunk 1]
```diff
@@ -114,9 +114,33 @@ KUBE_UP_AUTOMATIC_CLEANUP=${KUBE_UP_AUTOMATIC_CLEANUP:-false}
...
 NETWORK_PROVIDER="${NETWORK_PROVIDER:-none}" # opencontrail, flannel
-OPENCONTRAIL_TAG="${OPENCONTRAIL_TAG:-R2.20}"
-OPENCONTRAIL_KUBERNETES_TAG="${OPENCONTRAIL_KUBERNETES_TAG:-master}"
-OPENCONTRAIL_PUBLIC_SUBNET="${OPENCONTRAIL_PUBLIC_SUBNET:-10.1.0.0/16}"
 
 # Optional: if set to true, kube-up will configure the cluster to run e2e tests.
 E2E_STORAGE_TEST_ENVIRONMENT=${KUBE_E2E_STORAGE_TEST_ENVIRONMENT:-false}
+
+if [[ $NETWORK_PROVIDER == "opencontrail" ]]; then
+    img=$(gcloud compute images list | grep debian-7-backports | awk '{print $1}')
```
Comment (thockin) [inline cluster/gce/config-default.sh]: This feels like the tail wagging the dog.  Why does a network provider redefine all of these previous params?

These files are already way to complicated - this is guaranteed going to break.  Users will set flags up above and have them overridden here.   We (google) are not e2e testing this - what keeps it working?

File: cluster/saltbase/salt/docker/init.sls [hunk 2]
```diff
@@ -99,6 +99,24 @@ net.ipv4.ip_forward:
...
 
+# Install docker on debian wheezy 7 VM
+{% if grains.cloud is defined and grains.cloud == 'gce'
+   and grains.os_family == 'Debian'
+   and grains.oscodename == 'wheezy'
+   and not salt['file.file_exists']('/usr/bin/docker') -%}
+  docker-engine:
+    pkg.installed:
+     - sources:
+         - docker-engine_1.9.0: http://apt.dockerproject.org/repo/pool/main/d/docker-engine/docker-engine_1.9.0-0~wheezy_amd64.deb
```
Comment (thockin) [inline cluster/saltbase/salt/docker/init.sls]: We're already at 1.9.1 - this is a game of perpetual catchup

**Comment by thockin** [comment]
I'm not sure exactly what to do with this.  The footprint of it is WAY too big to really consider.  The GCE and AWS platforms are owned and maintained and this introduces a significant tax on that maintenance, for very little actual value (are there really users demanding to run contrail on GCE?)  It's a nice demo, but it seems more geared to on-prem.

This is just not surgical enough to merge into the primary platform that we test on.  We are ill equipped to handle this very well.  Can we find better ways of parameterizing?  What if this were a whole separate getting-started guide?

@mikedanese who is spending time thinking about bootstrapping
@erictune because we discussed (no action required)

File: cluster/gce/config-default.sh (hunk 1 above)
Comment (sanabby) [inline cluster/gce/config-default.sh]: Thanks for taking time to review. 

The params are not redefined. Its just defaulted to a value. 

About 5 rebases back, debian 7 wheezy backports was not working as the image itself was changed and did not have support for cgroup memory in the default grub. I see that updated now. 

Removed the default values for these now.

File: cluster/saltbase/salt/docker/init.sls (hunk 2 above)
Comment (sanabby) [inline cluster/saltbase/salt/docker/init.sls]: About 5 rebases back, debian 7 wheezy backports was not working as the image itself was changed and did not have support for cgroup memory in the default grub. I see that updated now.

Reset this file to master.  You will not see this in the PR now.

**Comment by ixdy** [comment]
@k8s-bot unit test this please

**Comment by fejta** [comment]
This PR has had no meaningful activity for multiple months. If it is still valid please rebase, push a new commit and reopen the PR. Thanks!

**Comment by v1k0d3n** [comment]
@sanabby any updates on this? it would be really nice to see this added, but no updates. do you have another PR?

**Comment by sanabby** [comment]
@v1k0d3n I am working on submitting a different PR after rebasing. Will keep you in the loop on the PR from the rebased branch so you can try using opencontrail network plugin.

**Comment by rantuttl** [comment]
@sanabby I'm with @v1k0d3n. Really want to give it a try!! Any ETA following rebase?

**Comment by thockin** [comment]
We're working on a plan that will make integrations like this WAY more
self-contained.

**Pull request closed without merging**